python/raft/common/handle.pyx

```diff
@@ -50,51 +51,36 @@ cdef class Handle:
         del handle  # optional!
     """
 
-    # handle_t doesn't have copy operator. So, use pointer for the object
-    # python world cannot access to this raw object directly, hence use
-    # 'size_t'!
-    cdef size_t h
-
-    # not using __dict__ unless we need it to keep this Extension as lean as
-    # possible
-    cdef int n_streams
-
-    def __cinit__(self, n_streams=0):
+    def __cinit__(self, stream=None, n_streams=0):
         self.n_streams = n_streams
-        self.h = <size_t>(new handle_t(n_streams))
-
-    def __dealloc__(self):
-        h_ = <handle_t*>self.h
-        del h_
-
-    def setStream(self, stream):
-        cdef size_t s = <size_t>stream.getStream()
-        cdef handle_t* h_ = <handle_t*>self.h
-        h_.set_stream(<_Stream>s)
-
-    def sync(self):
+        self.stream_pool.reset(new cuda_stream_pool(n_streams))
+
+        cdef cuda_stream_view c_stream
+        if stream is None:
+            # this constructor will construct a "main" handle on null stream
```

Comment:
While I can't imagine too many cases where we require asynchronicity in the Python layer currently, I would think we'd want to follow the same pattern that we're following in C++ and create a stream if it doesn't exist so that we aren't blocking off the entire GPU, right? For example, if someone is using multiple processes to submit jobs to the same GPU, the kernel launches won't be asynchronous wrt each other. Or was this discussed already and I'm just late to the game?

Reply:
This is a very, very nice catch. We weren't accounting for this up until now either. This doc should shed better light on this https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html.

In that case, we shouldn't use https://github.com/rapidsai/rmm/blob/2179ffd5dbc2847c00d2685c010b345f366f2816/include/rmm/cuda_stream_view.hpp#L114 or https://github.com/rapidsai/rmm/blob/2179ffd5dbc2847c00d2685c010b345f366f2816/include/rmm/cuda_stream_view.hpp#L119

We should instead use this as default: https://github.com/rapidsai/rmm/blob/2179ffd5dbc2847c00d2685c010b345f366f2816/include/rmm/cuda_stream_view.hpp#L124

Let me know if this solution looks correct